src/lib/Strategy/Storage/GroupByItemTypeAndLanguageStrategy.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace EzSystems\EzRecommendationClient\Strategy\Storage;
+
+use EzSystems\EzRecommendationClient\Criteria\Criteria;
+use EzSystems\EzRecommendationClient\Service\Storage\DataSourceServiceInterface;
+use EzSystems\EzRecommendationClient\Value\Storage\ItemGroup;
+use EzSystems\EzRecommendationClient\Value\Storage\ItemGroupList;
+use Ibexa\Contracts\Personalization\Criteria\CriteriaInterface;
+use Ibexa\Contracts\Personalization\Value\ItemGroupListInterface;
+
+final class GroupByItemTypeAndLanguageStrategy implements GroupItemStrategyInterface
+{
+    private DataSourceServiceInterface $dataSourceService;
```

Comment:
It doesn't feel right.

`DataSourceService` you declared depends on `GroupItemStrategyInterface`, and here this implementation depends on `DataSourceServiceInterface`. I tend to avoid circular references like that.

What's more I assume you always want to call the grouping strategy in context of the calling data source. I'd change the grouping interface methods to include the data source instead. This would lead to method calls like so:
```php
interface GroupItemStrategyInterface
{
    public function getGroupList(DataSourceServiceInterface $dataSource, CriteriaInterface $criteria): ItemGroupListInterface;
}

$this->groupingStrategy->getGroupList($this, ...);
```

Although for class design I think @alongosz is better suited, and should give his insight.